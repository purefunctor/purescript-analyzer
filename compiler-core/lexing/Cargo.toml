[package]
name = "lexing"
version = "0.1.0"
edition = "2021"

[dependencies]
memchr = "2.7.4"
syntax = { version = "0.1.0", path = "../syntax" }
unicode_categories = "0.1.1"

[dev-dependencies]
insta = "1.41.1"
test_each_file = "0.3.4"

