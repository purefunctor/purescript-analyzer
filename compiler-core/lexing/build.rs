fn main() {
    println!("cargo:rerun-if-changed=tests/layout");
    println!("cargo:rerun-if-changed=tests/lexer");
}
