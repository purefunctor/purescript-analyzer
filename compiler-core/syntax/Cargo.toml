[package]
name = "syntax"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
hashbrown = "0.15.2"
la-arena = "0.3.1"
paste = "1.0.15"
rowan = "0.16.1"
rustc-hash = "2.1.1"
