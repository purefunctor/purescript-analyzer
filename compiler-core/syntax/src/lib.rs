pub mod cst;

#[macro_use]
pub mod map;

mod token_set;

pub use map::AstPtrMap;
pub use token_set::TokenSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
#[allow(non_camel_case_types)]
pub enum SyntaxKind {
    // Generic
    TEXT = 0,

    // Names
    HOLE,
    UPPER,
    LOWER,
    OPERATOR,
    OPERATOR_NAME,

    // Operators
    AT,
    BACKSLASH,
    COLON,
    DOUBLE_COLON,
    COMMA,
    EQUAL,
    MINUS,
    PERIOD,
    DOUBLE_PERIOD,
    DOUBLE_PERIOD_OPERATOR_NAME,
    PIPE,
    TICK,
    UNDERSCORE,
    LEFT_ARROW,
    RIGHT_ARROW,
    LEFT_THICK_ARROW,
    RIGHT_THICK_ARROW,
    LEFT_PARENTHESIS,
    RIGHT_PARENTHESIS,
    LEFT_CURLY,
    RIGHT_CURLY,
    LEFT_SQUARE,
    RIGHT_SQUARE,

    // Keywords
    ADO,
    AS,
    CASE,
    CLASS,
    DATA,
    DERIVE,
    DO,
    ELSE,
    FORALL,
    FOREIGN,
    HIDING,
    IF,
    IMPORT,
    IN,
    INFIX,
    INFIXL,
    INFIXR,
    INSTANCE,
    LET,
    MODULE,
    NEWTYPE,
    NOMINAL,
    OF,
    PHANTOM,
    REPRESENTATIONAL,
    ROLE,
    THEN,
    TYPE,
    WHERE,

    // Literals
    CHAR,
    INTEGER,
    NUMBER,
    RAW_STRING,
    STRING,
    TRUE,
    FALSE,

    // Control
    LAYOUT_START,
    LAYOUT_SEPARATOR,
    LAYOUT_END,
    END_OF_FILE,

    // Special
    ERROR,

    // Nodes
    Node,
    Annotation,
    Qualifier,

    ModuleName,
    QualifiedName,
    LabelName,

    Module,
    ModuleHeader,
    ExportList,
    ExportValue,
    ExportClass,
    ExportType,
    ExportOperator,
    ExportTypeOperator,
    ExportModule,
    ModuleImports,
    ImportStatement,
    ImportList,
    ImportValue,
    ImportClass,
    ImportType,
    ImportOperator,
    ImportTypeOperator,
    ImportAlias,
    ModuleStatements,

    TypeItemsAll,
    TypeItemsList,

    ValueSignature,
    ValueEquation,
    FunctionBinders,

    Unconditional,
    WhereExpression,

    Conditionals,
    PatternGuarded,
    PatternGuardBinder,
    PatternGuardExpression,

    LetBindingSignature,
    LetBindingEquation,
    LetBindingPattern,
    LetBindingStatements,

    TypeApplicationChain,
    TypeArrow,
    TypeConstrained,
    TypeConstructor,
    TypeForall,
    TypeHole,
    TypeInteger,
    TypeKinded,
    TypeOperator,
    TypeOperatorChain,
    TypeOperatorPair,
    TypeString,
    TypeVariable,
    TypeVariableBinding,
    TypeWildcard,
    TypeRecord,
    TypeRow,
    TypeRowItem,
    TypeRowTail,
    TypeParenthesized,

    BinderTyped,
    BinderOperatorChain,
    BinderOperatorPair,
    BinderInteger,
    BinderNumber,
    BinderConstructor,
    BinderVariable,
    BinderNamed,
    BinderWildcard,
    BinderString,
    BinderChar,
    BinderTrue,
    BinderFalse,
    BinderArray,
    BinderRecord,
    BinderParenthesized,

    ExpressionTyped,
    ExpressionOperatorChain,
    ExpressionOperatorPair,
    ExpressionInfixChain,
    ExpressionInfixPair,
    ExpressionTick,
    ExpressionNegate,
    ExpressionApplicationChain,
    ExpressionTypeArgument,
    ExpressionTermArgument,
    ExpressionIfThenElse,
    ExpressionIf,
    ExpressionThen,
    ExpressionElse,
    ExpressionLetIn,
    ExpressionLambda,
    ExpressionCaseOf,
    ExpressionDo,
    ExpressionAdo,
    ExpressionConstructor,
    ExpressionVariable,
    ExpressionOperatorName,
    ExpressionSection,
    ExpressionHole,
    ExpressionString,
    ExpressionChar,
    ExpressionTrue,
    ExpressionFalse,
    ExpressionInteger,
    ExpressionNumber,
    ExpressionArray,
    ExpressionRecord,
    ExpressionParenthesized,
    ExpressionRecordAccess,
    ExpressionRecordUpdate,

    CaseTrunk,
    CaseBranches,
    CaseBranchBinders,
    CaseBranch,

    DoStatements,
    DoStatementBind,
    DoStatementLet,
    DoStatementDiscard,

    RecordField,
    RecordPun,
    RecordUpdates,
    RecordUpdateLeaf,
    RecordUpdateBranch,

    InfixDeclaration,

    TypeRoleDeclaration,
    TypeRole,

    TypeSynonymSignature,
    TypeSynonymEquation,

    ClassSignature,
    ClassDeclaration,
    ClassConstraints,
    ClassHead,
    ClassFunctionalDependencies,
    FunctionalDependencyDetermined,
    FunctionalDependencyDetermines,
    ClassStatements,
    ClassMemberStatement,

    InstanceChain,
    InstanceDeclaration,
    InstanceName,
    InstanceConstraints,
    InstanceHead,
    InstanceStatements,
    InstanceSignatureStatement,
    InstanceEquationStatement,

    ForeignImportDataDeclaration,
    ForeignImportValueDeclaration,

    NewtypeSignature,
    NewtypeEquation,

    DataSignature,
    DataEquation,
    DataConstructor,

    DeriveDeclaration,

    #[doc(hidden)]
    __LAST,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PureScript {}

impl rowan::Language for PureScript {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::__LAST as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind as u16)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(value: SyntaxKind) -> Self {
        Self(value as u16)
    }
}

pub type SyntaxNode = rowan::SyntaxNode<PureScript>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<PureScript>;
pub type SyntaxNodePtr = rowan::ast::SyntaxNodePtr<PureScript>;
pub type SyntaxToken = rowan::SyntaxToken<PureScript>;
pub type SyntaxElement = rowan::SyntaxElement<PureScript>;

impl SyntaxKind {
    pub fn is_layout_token(&self) -> bool {
        matches!(self, Self::LAYOUT_START | Self::LAYOUT_SEPARATOR | Self::LAYOUT_END)
    }
}
