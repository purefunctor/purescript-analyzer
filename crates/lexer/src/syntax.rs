#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum SyntaxKind {
    // region: whitespace
    WHITESPACE,
    LINE_COMMENT,
    BLOCK_COMMENT,
    // endregion: whitespace

    // region: keywords
    /// _
    UNDERSCORE,
    /// ado
    ADO,
    /// as
    AS,
    /// case
    CASE,
    /// class
    CLASS,
    /// data
    DATA,
    /// derive
    DERIVE,
    /// do
    DO,
    /// else
    ELSE,
    /// false
    FALSE,
    /// foreign
    FOREIGN,
    /// hiding
    HIDING,
    /// if
    IF,
    /// import
    IMPORT,
    /// in
    IN,
    /// infix
    INFIX,
    /// infixl
    INFIXL,
    /// infixr
    INFIXR,
    /// instance
    INSTANCE,
    /// let
    LET,
    /// module
    MODULE,
    /// newtype
    NEWTYPE,
    /// of
    OF,
    /// then
    THEN,
    /// true
    TRUE,
    /// type
    TYPE,
    /// where
    WHERE,
    // endregion: keywords

    // region: names
    /// Module names
    MODULE_NAME,
    /// Named holes
    SOURCE_HOLE,
    /// Lower-case name
    SOURCE_IDENTIFIER,
    /// Upper-case name
    SOURCE_PROPER,
    /// +, -, *, /
    SOURCE_OPERATOR,
    /// (+), (-), (*), (/)
    SOURCE_SYMBOL,
    // endregion: names

    // region: operators
    /// <-, ←
    LEFT_ARROW,
    /// <=, ⇐
    LEFT_DOUBLE_ARROW,
    /// ->, →
    RIGHT_ARROW,
    /// =>, ⇒
    RIGHT_DOUBLE_ARROW,
    /// ::, ∷
    DOUBLE_COLON,
    /// forall, ∀
    FORALL,
    /// =
    EQUALS,
    /// .
    DOT,
    /// ,
    COMMA,
    /// \\
    BACKSLASH,
    /// |
    PIPE,
    /// @
    AT,
    /// `
    TICK,
    // endregion: operators

    // region: delimiters
    /// (
    LEFT_PARENTHESIS,
    /// )
    RIGHT_PARENTHESIS,
    /// [
    LEFT_SQUARE,
    /// ]
    RIGHT_SQUARE,
    /// {
    LEFT_CURLY,
    /// }
    RIGHT_CURLY,
    // endregion: delimiters

    // region: values
    /// ""
    STRING,
    /// """
    RAW_STRING,
    /// 'a', '\n', '\xFFFFFF'
    CHAR,
    /// 0, 0xFF
    INTEGER,
    /// 0.0, 1.23e+5
    NUMBER,
    // endregion: values

    // region: control
    END_OF_FILE,
    LAYOUT_START,
    LAYOUT_DELIMITER,
    LAYOUT_END,
    // endregion: control
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PureScript {}

impl rowan::Language for PureScript {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::END_OF_FILE as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind as u16)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(value: SyntaxKind) -> Self {
        Self(value as u16)
    }
}

pub type SyntaxNode = rowan::SyntaxNode<PureScript>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<PureScript>;
pub type SyntaxNodePtr = rowan::ast::SyntaxNodePtr<PureScript>;
pub type SyntaxToken = rowan::SyntaxToken<PureScript>;
pub type SyntaxElement = rowan::SyntaxElement<PureScript>;

impl SyntaxKind {
    pub fn is_whitespace_or_comment(self) -> bool {
        matches!(
            self,
            SyntaxKind::WHITESPACE | SyntaxKind::LINE_COMMENT | SyntaxKind::BLOCK_COMMENT
        )
    }
}
