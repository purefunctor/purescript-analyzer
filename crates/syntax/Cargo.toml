[package]
name = "syntax"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rowan = "0.15.15"
smol_str = "0.2.0"
