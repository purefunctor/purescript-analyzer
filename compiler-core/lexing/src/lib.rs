mod categories;
mod layout;
mod lexed;
mod lexer;

pub use lexed::Lexed;
use syntax::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Position {
    pub line: u32,
    pub column: u32,
}

pub fn lex(source: &str) -> Lexed<'_> {
    let mut lexer = lexer::Lexer::new(source);
    while !lexer.is_eof() {
        lexer.take_token();
    }
    lexer.finish()
}

pub fn layout(lexed: &Lexed) -> Vec<SyntaxKind> {
    let mut layout = layout::Layout::new(lexed);
    while !layout.is_eof() {
        layout.take_token();
    }
    layout.finish()
}
