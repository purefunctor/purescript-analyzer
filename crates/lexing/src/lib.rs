//! A lexer for PureScript source code.

mod layout;
mod lexed;
mod lexer;

pub use lexed::Lexed;

use layout::Machine;
use lexer::Lexer;
use syntax::SyntaxKind;

/// Tokenizes a source string.
pub fn lex(source: &str) -> Lexed {
    let mut lexer = Lexer::new(source);
    loop {
        if lexer.is_eof() {
            break lexer.finalize();
        }
        lexer.take_token();
    }
}

/// Applies the layout algorithm.
pub fn layout(lexed: &Lexed) -> Vec<SyntaxKind> {
    let mut machine = Machine::new(lexed);
    loop {
        if machine.is_eof() {
            break machine.finalize();
        }
        machine.take_token();
    }
}

#[cfg(test)]
mod tests {
    use syntax::SyntaxKind;

    use crate::{layout, lex};

    const FULL_SOURCE: &str = r"module Test where

test = ado
    baz
    let foo = bar
    in bar

test = ado in foo

test = ado
    foo <- bar $ let a = 42 in a
    baz <- b
    in bar

test = ado
    foo
    let bar = let a = 42 in a
    let baz = 42
    in bar
    
example1 = do
  foo bar
  <|> baz

example2 = do
  foo bar
  `wat` baz

example3 =
  case _ of
    Foo a -> 1
    Bar b -> 2
    `append` 3

example4 =
  case _ of
    Foo a -> 1
    Bar b -> 2
    + 3

-- Including data because of `|` masking
data Foo
 = Foo
 | Bar
 | Baz

test =
  case foo of
    a | b, c ->
      d
    a | b, c -> d

test = case a, b of
  c, d
   | e ->
     case e of
       f | true -> bar
         | false -> baz
   | f -> g

test a
  | false =
      case false of
        true | a > 12 -> true
  | otherwise = true

test = case a of foo | foo \a -> a -> true

test = a `case _ of x | unit # \_ -> true, true -> const` b

test = case a of
  12 | do that
          that   -> this
     | otherwise -> this

test a b = [ case _ of
  12 | case a, b of
         _, 42 -> b
         _, 12 -> false, b -> true
     | case a, b of
         _, 42 -> b
         _, 12 -> false, b -> true, false ]

test a
  | case a, b of
         _, 42 -> b
         _, 12 -> false, b = true
  | case a, b of
         _, 42 -> b
         _, 12 -> false, b = true

test = case foo of
  Nothing -> a
    where a = 12
  Just a -> do
    what
  where
    foo = bar

test = case f of Foo -> do that
                        where foo = 12

import Foo (class Foo)

class Foo a b c d | a -> b, c -> d where
  foo :: Foo

class Foo a b c d | a -> b, c -> d

instance foo :: Foo

test =
  [ case do foo, bar of
      a | b, c -> d, bar
    ]

test =
  [ case do foo, bar of a | b, c -> d, bar ]

test =
  [ do do do foo, bar ]

test =
  [ \foo -> foo, bar ]

test = foo where
  bar =
    case a, b of
      c, d | d == [case true, w of 1, a -> true, false ] -> d
      e, d | do what, do that -> d

test1 = a
test2 = {
  b
}
test3 = do
  foo
  bar (
    baz
  ) == 12
  baz
test4 = c

test = do
  let foo = bar
  foo

test = do
  let foo = bar
  in baz
  foo

test = do
  let foo = bar
    in baz
  foo

test = do
  foo
  foo do
    bar
  <|> bar

test =
  do
    do do
        foo where bar = baz

foo = do
  if true then
    false
  else if false then do
    that
    else do
     what
  that

instance foo :: Foo Int else bar :: Foo String
else baz :: Foo Boolean

type IntType = (-1)

type IntType' = (-
  -- here's a comment
  1)

test =
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  in
    foo

test = do
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  foo

test = ado
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  foo";

    #[test]
    fn tokens_are_identical() {
        let lexed = lex(FULL_SOURCE);

        let mut left = vec![];
        for index in 0..lexed.len() {
            let kind = lexed.kind(index);
            if kind.is_whitespace_or_comment() {
                continue;
            }
            left.push(kind);
        }

        let right: Vec<_> = layout(&lexed)
            .into_iter()
            .filter(|kind| {
                !matches!(
                    kind,
                    SyntaxKind::LayoutStart | SyntaxKind::LayoutSep | SyntaxKind::LayoutEnd
                )
            })
            .collect();

        assert_eq!(left, right);
    }
}
