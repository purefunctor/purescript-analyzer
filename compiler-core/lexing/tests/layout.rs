use lexing::Lexed;
use syntax::SyntaxKind;
use test_each_file::test_each_file;

fn print_tokens(lexed: &Lexed, tokens: &[SyntaxKind]) -> String {
    let mut buffer = String::new();
    let mut index = 0;

    for token in tokens {
        match token {
            SyntaxKind::LAYOUT_START => {
                buffer.push('{');
            }
            SyntaxKind::LAYOUT_SEPARATOR => {
                buffer.push(';');
            }
            SyntaxKind::LAYOUT_END => {
                buffer.push('}');
            }
            i => {
                if let Some(annotation) = lexed.annotation(index) {
                    buffer.push_str(annotation);
                }
                if let Some(qualifier) = lexed.qualifier(index) {
                    buffer.push_str(qualifier);
                }
                if matches!(i, SyntaxKind::END_OF_FILE) {
                    buffer.push_str("<eof>");
                } else {
                    buffer.push_str(lexed.text(index));
                }
                index += 1;
            }
        }
    }
    buffer
}

test_each_file! { in "./compiler-core/lexing/tests/layout" => |content: &str| {
    let lexed = lexing::lex(content);
    let tokens = lexing::layout(&lexed);
    insta::assert_snapshot!(print_tokens(&lexed, &tokens));
}}
