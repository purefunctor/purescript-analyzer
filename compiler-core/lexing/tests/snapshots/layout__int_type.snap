---
source: compiler-core/lexing/tests/layout.rs
expression: source
snapshot_kind: text
---
module Test where{

type IntType = (-1);

type IntType' = (-
  -- here's a comment
  1)}
<eof>
